Add a delete tag confirmation

sandymcfadden opened the pull request:
### Fix

Currently, when deleting a tag if you click the delete icon it immediately is deleted. This PR resolves #2568 by adding a confirmation dialog where you must confirm the deletion.

<img width="944" alt="Screen Shot 2021-03-04 at 7 37 48 AM" src="https://user-images.githubusercontent.com/1326294/109958492-a8a4ca80-7cbc-11eb-8af2-80218ddc6458.png">
<img width="940" alt="Screen Shot 2021-03-04 at 7 38 12 AM" src="https://user-images.githubusercontent.com/1326294/109958509-ac385180-7cbc-11eb-8829-43599cf42c96.png">


### Test

1. Test in both light and dark mode.
2. Open tag list.
3. Click to edit the tags.
4. Click on the trash icon to delete a tag.
5. Notice the dialog appears asking for confirmation.
6. Click the delete button.
7. Notice the tag is deleted.
8. Follow through tests again, but instead of clicking the delete button, close the dialog.
9. Notice the tag isn't deleted.

### Release

- Adds a confirmation dialog before deleting a tag.

dmsnell commented on lib/types.ts:
```diff
@@ -66,7 +66,8 @@ export type DialogType =
   | 'SETTINGS'
-  | 'SHARE';
+  | 'SHARE'
+  | 'TRASH-TAG-CONFIRMATION';
```
nice work finding this and adding the new dialog to the type 👍

dmsnell commented on lib/dialogs/trash-tag-confirmation/index.tsx:
```diff
@@ -0,0 +1,60 @@
+import actions from 'lib/state/actions';
```
it probably feels arbitrary but this, for consistency sake, belongs below third-party imports. it looks like your IDE might have auto-included this and the imports in line 5. we can replace both import statements with the more standard-in-this-app `import actions from '../../state/actions'` and then call using the fully-qualified action access, like `actions.ui.closeDialog`

sandymcfadden replied:
Thanks for catching this!

dmsnell commented on lib/dialogs/trash-tag-confirmation/index.tsx:
```diff
@@ -0,0 +1,60 @@
+      className="button-primary delete-tag"
+      onClick={() => {
+        trashTag(tagToTrash);
+        closeDialog();
```
we want to avoid this kind of "double-dispatch" because it introduces _partial updates_ to our app, where the UI state is in transition. much better is to look at the dialog reducer and ask ourselves, "what changes this data?" we can answer that by saying, "if we trash a tag, we imply that we should close the tag-trash-confirmation dialog."

in that case we only need to dispatch the single action which expresses the user-oriented intent "yes, trash this tag," and let the application _react_ to that however it wants.

from time to time we may only want to conditionally do something like this, like close the dialog on trashing a tag. when those cases arise we might want to include metadata on the Redux action to indicate to the appropriate reducer that yes indeed this implies closing the dialog. here I'm not sure that extra metadata would be warranted because why would we continue showing that confirmation dialog if the tag was trashed.

that raises another point concerning the dialog reducer. _if_ it's open and the tag is trashed from another device we might want to close the dialog, or if we want to keep it open then we could have the dialog change to reflect that the tag was just remotely trashed

dmsnell commented on lib/state/ui/middleware.ts:
```diff
@@ -91,6 +93,10 @@ export const middleware: S.Middleware = (store) => (
 
+    case 'TAG_TO_TRASH':
+      store.dispatch(showDialog('TRASH-TAG-CONFIRMATION'));
+      return next(action);
```
we should not need this. although as we think about it we think, "when we want to trash a tag then open the dialog"  but the dialog reducer can listen and react on its own without a second dispatch. ping me in Slack if you want to talk through this

dmsnell reviewed (commented):
looking good @sandymcfadden 

I'm noting how almost all of the places we open a dialog it's purely a dialog name we need. I wonder if we can leave `actions.ui.showDialog` as it was, or add an optional second parameter that only `trashTag` need supply? that would leave all of the other existing calls unchanged, and the same could apply to `appCommandSender`, since it doesn't (currently) trash a tag. might be the case where if we need the additional data we could find a way to do it without exposing the internals of the action to the callers, as we can see is done when we require them to pass `{type: ""}` vs. just the name.

sandymcfadden commented:
Thank you @dmsnell. I've updated things based on the feedback, but still don't have the types all set correctly as I'm getting errors in VS Code. Things all do seem to be working though.

dmsnell commented on lib/state/action-types.ts:
```diff
@@ -101,7 +101,13 @@ export type SetUnsyncedNoteIds = Action<
+  'SHOW_DIALOG',
+  {
+    dialogName: T.DialogType['type'];
+    dialogData: object;
```
generally speaking we can omit the prefix-as-namespace since the action type itself indicates we are talking about a dialog. in the previous code we had to have something there to indicate which one, and we probably could have used `name` but in the absence of a clear word for it, `dialog` communicated "which dialog"

here we can probably communicate just as effectively if we store `name` and `data`

sandymcfadden replied:
Sounds good and makes sense to me.

dmsnell commented on lib/state/ui/reducer.ts:
```diff
@@ -90,9 +90,20 @@ const dialogs: A.Reducer<T.DialogType[]> = (state = [], action) => {
+    case 'TRASH_TAG':
+      return state.filter(
+        (element) => element.type !== 'TRASH-TAG-CONFIRMATION'
+      );
```
do you have `prettier` and the git hooks setup? I thought it would have added a newline between the `case` lines

sandymcfadden replied:
I believe I have it all set up properly, but it doesn't add a newline between the `case` lines. 
It looks like there is a mix of styles used after a quick search where in some places there are spaces but not everywhere.

dmsnell replied:
interesting. thanks for checking

dmsnell commented on lib/state/ui/reducer.ts:
```diff
@@ -90,9 +90,20 @@ const dialogs: A.Reducer<T.DialogType[]> = (state = [], action) => {
+    case 'REMOTE_TAG_DELETE':
+      return state.filter(
+        (element) =>
+          element.tagName !== tagHashOf(action.tagHash) ||
```
I'm a bit lost in the negative-or logic with the filter. would it be clearer if we stated the specific dialog we want to get rid of instead of the qualifications for which ones can stay?

```ts
return state.filter(element => !(
	element.type === 'TRASH-TAG-CONFIRMATION' &&
	element.tagName === tagHashOf(action.tagHash)
))
```

sandymcfadden replied:
Thank you! I was confusing myself with this logic as well. This suggestion is so much more readable.

dmsnell commented on lib/app-layout/index.tsx:
```diff
@@ -144,7 +144,9 @@ export class AppLayout extends Component<Props> {
-  keyboardShortcutsAreOpen: state.ui.dialogs.includes('KEYBINDINGS'),
+  keyboardShortcutsAreOpen:
+    state.ui.dialogs.find((element) => element.type === 'KEYBINDINGS') !==
+    undefined,
```
simply because this is a big complicated and specific to the dialogs reducer I wonder if we should add a selector in `selectors.ts` to make it clearer in the normal code:

```ts
export const isDialogOpen = (state: S.State, name: T.DialogType) =>
	state.ui.dialogs.find(({type}) => type === name) !== undefined
```

```ts
import * from selectors;

…

keyboardShortcutsAreOpen: selectors.isDialogOpen(state, 'KEYBINDINGS'),
```

normally I wouldn't encourage writing a selector or function for a single use like this but that's kind of quirky code _necessary_ to get the right answer and moving it into its own testable place might be helpful

sandymcfadden replied:
This certainly makes it cleaner to read when in `app-layout.tsx` I end up with a type error for `'KEYBINDINGS'` though `Argument of type 'string' is not assignable to parameter of type 'DialogType'.ts(2345)`

There are a couple of other places where I've run into similar as well. I'm not sure how to best deal with it. Another example is in the [reducer](https://github.com/Automattic/simplenote-electron/blob/add/delete-tag-confirmation-middleware/lib/state/ui/reducer.ts#L108). The last `tagName` there gives the error 
`Property 'tagName' does not exist on type 'Action<"SHOW_DIALOG", { name: "ABOUT" | "BETA-WARNING" | "CLOSE-WINDOW-CONFIRMATION" | "IMPORT" | "KEYBINDINGS" | "LOGOUT-CONFIRMATION" | "SETTINGS" | "SHARE" | "TRASH-TAG-CONFIRMATION"; data: object; }>'.ts(2339)`

It does all work though, any ideas or pointers you have?

dmsnell replied:
well my code examples are rarely correct or without bugs. `T.DialogType` was updated in this PR and I was reusing the old type. you can get just the `type` name by using `T.DialogType['type']`

sandymcfadden replied:
Of course, I totally missed that.

dmsnell commented on lib/dialog-renderer/index.tsx:
```diff
@@ -36,35 +37,39 @@ export class DialogRenderer extends Component<Props> {
-
     return (
       <Fragment>
         {this.props.dialogs.map((dialog) => (
```
is this happy with the types? does it know that `tagName` exists below? sometimes it has problems with `map()`

dmsnell replied:
also you can add `dialogs` above in line 38 and then refer to it here without `this.props` - looks like that was something that already existed, and maybe there's a reason (shadowing?) to avoid `dialogs` but if not you can it

sandymcfadden replied:
Types seem to be fine here. 
I added `dialogs` as you mentioned and that works fine, thank you.

dmsnell commented on lib/state/ui/reducer.ts:
```diff
@@ -90,9 +90,22 @@ const dialogs: A.Reducer<T.DialogType[]> = (state = [], action) => {
-
+    case 'TRASH_TAG':
+      return state.filter(
+        (element) => element.type !== 'TRASH-TAG-CONFIRMATION'
```
at this point I believe we are dealing with `dialog`s - might be more helpful than the generic "element"

sandymcfadden replied:
Very true. I'll update this.

dmsnell commented on lib/state/ui/reducer.ts:
```diff
@@ -90,9 +90,22 @@ const dialogs: A.Reducer<T.DialogType[]> = (state = [], action) => {
+      );
     case 'SHOW_DIALOG':
-      return state.includes(action.dialog) ? state : [...state, action.dialog];
+      return state.find((element) => element.type === action.name)
```
this prevents overlapping multiple tag trash dialogs over each other, which I believe is the right move anyway - just noting it. might be worth a comment in the code to tell future us that it's not a mistake

sandymcfadden replied:
Good call. 
I do have a question about this section @dmsnell 

`return state.find((dialog) => dialog.type === action.name)
        ? state
        : [...state, { type: action.name, tagName: action.tagName }];`

Is there some way I can make this more generic? I tried a few things but couldn't get it to work.
If we add more dialogs that have extra data it would have to be manually added here. 
I figure there should be a way to do it because in the `ShowDialog` [type](https://github.com/Automattic/simplenote-electron/blob/add/delete-tag-confirmation-middleware/lib/state/action-types.ts#L108) we are using the generic `data` property. But I can't seem to figure out how to get this to work.

dmsnell replied:
I don't know how much more generic we want it. that's a question at least this PR doesn't raise: the generic question would likely be "can we have more than one dialog of a specific type open if they have different data?" for tag deletion it seems reasonable to not allow that, and we don't currently have any other dialogs with associated data.

were you thinking of another kind of generalization?

sandymcfadden replied:
You got at what I was looking for in your other comment. Thank you!

dmsnell reviewed (commented):
Left a few final observations you can choose to respond to; you don't need to take them all but I think those are the only things that I see as being considerations before a merge. Otherwise the code looks good and assuming you have tested it then it looks good to me. I did not test this.

dmsnell commented on lib/state/ui/reducer.ts:
```diff
@@ -90,9 +90,21 @@ const dialogs: A.Reducer<T.DialogType[]> = (state = [], action) => {
+      // This ensures we only show one dialog of each type at a time.
+      return state.find((dialog) => dialog.type === action.name)
+        ? state
+        : [...state, { type: action.name, tagName: action.tagName }];
```
we don't need `tagName` called out specifically here - maybe I misread your comment about generalization.

note that we have access to associated data via the action. we can accomplish this in a couple ways, but if we want to start by blindly accepting each additional bit of data, we can spread the rest of the arguments.

```ts
case 'SHOW_DIALOG': {
	const {type, name, meta, ...data} = action;

	…
		? state
		: [...state, {type: action.name, ...data}];
}
```

sandymcfadden replied:
This is exactly what I was trying to get at and figure out in my other comment, thank you!

belcherj commented:
Be sure to tag @SylvesterWilmott for design review.

belcherj commented:
I think the corners are more rounded in the design.

<img width="391" alt="Screen Shot 2021-03-01 at 12 13 04 PM" src="https://user-images.githubusercontent.com/6817400/109532812-8118ec80-7a87-11eb-9f0c-7fc74c299b85.png">
<img width="392" alt="Screen Shot 2021-03-01 at 12 12 42 PM" src="https://user-images.githubusercontent.com/6817400/109532814-8118ec80-7a87-11eb-8986-ec9c75a8cae2.png">

sandymcfadden commented:
You're absolutely right @belcherj. I've fixed that now. It does still need design review though as I added the tag name to the text of the dialog which wasn't included in the design. I wanted to get your opinion on that @SylvesterWilmott 
![Screen Shot 2021-03-01 at 2 14 01 PM](https://user-images.githubusercontent.com/1326294/109540105-6e5ae380-7a98-11eb-88d3-1b6aca91d98e.png)

dmsnell commented on lib/dialogs/trash-tag-confirmation/index.tsx:
```diff
@@ -28,7 +28,9 @@ const TrashTagConfirmation: FunctionComponent<Props> = ({
   >
-    <p>Are you sure you want to delete &quot;{tagName}&quot;?</p>
+    <div className="dialog-inner-content">
+      Are you sure you want to delete &quot;{tagName}&quot;?
```
if it's giving you a headache with the formatting you can also produce the entire string in JavaScript

```ts
<div className="dialog-inner-content">{`Are you sure you want to delete '${tagName}'?`</div>
```

strings like this will be properly escaped by React for HTML

sandymcfadden replied:
Thanks. For once this is something I learned already :-)
The changes here were more about getting the spacing to be consistent with the designs.
I used the `&quot;` characters following another example I had found in the codebase.
Do you prefer one way over the other?

dmsnell replied:
no preference from me!

SylvesterWilmott commented:
After some back and forth on slack, the design looks good!

codebykat commented:
This is looking good.

I did have to test with this 😂

<img width="476" alt="Screen Shot 2021-03-03 at 11 43 37 PM" src="https://user-images.githubusercontent.com/52152/109928495-5cd13180-7c7a-11eb-8775-08a377f3a708.png">

Similar to the fix we need for the email verification screen, I bet. `word-break: break-all` ONLY on the tag name itself. #2712

sandymcfadden commented:
Thanks, @codebykat I have that fixed now.

codebykat reviewed (approved):
LGTM 👍

dmsnell reviewed (approved):
:shipit:

sandymcfadden committed:
Rebasing on develop before merging
